compiler/rustc_middle/src/ty/error.rs

```diff
@@ -93,20 +92,16 @@ impl TypeError<'_> {
 /// in parentheses after some larger message. You should also invoke `note_and_explain_type_err()`
 /// afterwards to present additional details, particularly when it comes to lifetime-related
 /// errors.
-impl<'tcx> fmt::Display for TypeError<'tcx> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+impl<'tcx> TypeError<'tcx> {
+    pub fn to_string(self, tcx: TyCtxt<'tcx>) -> String {
```

Comment:
Given that this is always in the error path, allocating a `String` is probably fine, though I think it would probably not be too onerous to retain the more optimal `fmt` approach by having a structure that combines `TyCtxt` with `TypeError`, somewhat like this:

```rust
impl<'tcx> TypeError<'tcx> {
    fn with(self, tcx: TyCtxt<'tcx>) -> TypeErrorWithTyCtxt<'tcx> { todo!() }
}

impl<'tcx> std::fmt::Display for TypeErrorWithTyCtxt<'tcx> { ... }
```

It would probably be terser, and I think we use this kind of pattern elsewhere in the codebase already, so it isn’t that unconventional.

Reply:
Maybe turning this to be Cow would be good enough?